library/manageiq_custom_attributes.py

```diff
@@ -0,0 +1,259 @@
+                    id=entity_id)
+            result = self.client.post(url, action='add', resources=custom_attributes)
+            self.changed = True
+            return result['results']
+        except Exception as e:
+            self.module.fail_json(msg="Failed to add the custom attributes. Error: {}".format(e))
+
+    def update_custom_attribute(self, entity_type, entity_id, ca, ca_href):
+        """ Returns the updated custom attributes """
+        try:
+            url = '{api_url}/{entity_type}/{id}/custom_attributes'.format(
+                    api_url=self.api_url,
+                    entity_type=ManageIQCustomAttributes.supported_entities[entity_type],
+                    id=entity_id)
+            ca_object = {'name': ca['name'], 'href': ca_href, 'value': ca['value']}
+            result = self.client.post(url, action='edit', resources=[ca_object])
+            self.changed = True
+            return result['results']
+        except Exception as e:
+            self.module.fail_json(msg="Failed to update the custom attribute {ca_name}. Error: {error}".format(ca_name=ca['name'], error=e))
+
+    def add_or_update_custom_attributes(self, entity_type, entity_name, custom_attributes):
+        """ Adds custom attributes to an entity in manageiq or updates the
+        attributes in case already exists
+
+        Returns:
+            the added or updated custom attributes, whether or not a change
+            took place and a short message describing the operation executed
+        """
+        added, updated = [], []
+        message = ""
+        # check if entity with the type and name passed exists in manageiq
+        entity_id = self.find_entity_by_name(entity_type, entity_name)
+        if not entity_id:  # entity doesn't exist
+            self.module.fail_json(
+                msg="Failed to add the custom attributes. {entity_type} {entity_name} does not exist".format(entity_type=entity_type, entity_name=entity_name))
+
+        entity_cas = self.get_entity_cas(entity_type, entity_id)
+        if not entity_cas:  # no custom attributes for this entity
+            added.extend(self.add_custom_attributes(entity_type, entity_id, custom_attributes))
+        else:
+            for new_ca in custom_attributes:
+                existing_ca = next((ca for ca in entity_cas if ca['name'] == new_ca['name']), None)
+                if existing_ca:
+                    if new_ca['value'] != existing_ca['value']:
+                        updated.extend(self.update_custom_attribute(entity_type, entity_id, new_ca, existing_ca['href']))
+                else:
+                    added.extend(self.add_custom_attributes(entity_type, entity_id, [new_ca]))
+
+        if added:
+            message = "Successfully added the custom attributes to {entity_name} {entity_type}"
+        if updated:
+            message += "Successfully updated the custom attributes of {entity_name} {entity_type}"
```

Comment:
If both happen, you'd get "..._type}Successfully".  At least add trailing dot.
Or just say "Successfully added/updated" or "Successfully set"; you return detailed `updates` dict anyway.